src/Model/Item.php

```diff
@@ -1553,9 +1553,10 @@ private static function getParentData(array $item)
 				Logger::info('tagged thread as mention', ['parent' => $item['parent'], 'uid' => $item['uid']]);
 			}
 
-			// Update the contact relations
+			// Update the contact anf gcontact relations
 			if ($item['author-id'] != $parent['author-id']) {
 				DBA::update('contact-relation', ['last-interaction' => $item['created']], ['cid' => $parent['author-id'], 'relation-cid' => $item['author-id']], true);
+				GContact::setRelation($parent['author-link'], $item['author-link'], $item['created']);
```

Comment:
This is insane to me. We should have only one `contact` table for all remote or local contacts, one `user-contact` table for the relationships between local users and contacts (follow direction, block, personal notes, custom display name and picture, etc...) and one `contact-relation` table for inter-contact relationships (follow direction, last interaction direction).

These three tables would replace:
- `contact` (split between `contact` with the public contact info and `user-contact` with the user contact info)
- `fcontact`
- `gcontact`
- `gcontact-relation`
- `gcign`?
- `gfollower`
- `glink`

I won't accept the creation of yet another table when we can and should shrink their number.

Reply:
"glink" will be removed in the near future. "apcontact" (ActivityPub) and "fcontact" (Diaspora) are protocol specific tables. "gcontact" are all contacts that we know of, they don't need to have interacted on this machine at all. "contact" are all contacts who were visible on this system, because they commented, posted or liked some content. "gcign" is some old stuff from the suggestions.

Reply:
Even then, `gcontact` and `contact` could easily be merged, same as `gfollower` and `gcontact-relation`.

Agreed to keep protocol-specific tables.

Do I need to provide a prototype of what I'm thinking about?

Reply:
`gfollower` and `gcontact-relation` are not the same and should not be mixed. The problem with `contact` and `gcontact` is that `contact` additionally contains the paths to the locally stored copies of the avatars. So a single `contact` entry always uses much more space than a single `gcontact` entry.

I have plans for a complete new structure - but this does require much more preparation. `gcontact-relation` and `contact-relation` are somehow redundant, so if you think so, I can just keep `contact-relation` and `gfollower`.

Reply:
> gfollower and gcontact-relation are not the same and should not be mixed.

Why not?

> The problem with contact and gcontact is that contact additionally contains the paths to the locally stored copies of the avatars.

Why not use the local proxy feature for this? When we save any contact, we can proxy remote contacts pictures while we store the avatar URL for local contacts. No more paths.

Reply:
`gfollower`is the exact representation of the "follower" and "following" endpoints from AP. This table is ideal for displaying which person follows which persons. But these needn't to be the people who have interacted. `contact-relation` is more suitable for contact suggestions since you can suggest contacts that are actively interacting with your contacts.The one table is information that we have to pull (so we need some `updated` field to see how old the data is) while the other table contains information that is pushed to us (and thus needs a field that shows the date of the last interaction). Trying to mix this, would end up in a table containing multiple date fields and multiple integer fields defining the type of relation between these contacts. This makes stuff unnecessarily complicate there. And in fact this wouldn't save us disk space as well.

Concerning the avatars I wouldn't want to use the proxy for that. We already have got our experience with this. Servers go offline all the time or are slow. So when we would proxy the pictures when - for example - a user was looking at their list of contacts then it would reduce the page load speed and would also leave unloaded pictures there for the offline servers. There is however a solution for this: We could leave the fields for the locally stored pictures `null` when creating the contact. And only when the contact is interacting (or when it is some user contact) we would fill the other fields as well. This would make the `gcontact` table obsolete.

Reply:
> Trying to mix this, would end up in a table containing multiple date fields and multiple integer fields defining the type of relation between these contacts. This makes stuff unnecessarily complicate there. And in fact this wouldn't save us disk space as well.

For me, it's simpler to have a single table with the same primary key. The primary keys of `gfollower` and `contact-relation` are both `contact.id as actor, contact.id as target` once we have a single `contact` table. I'm not considering disk space at all.

So the first step of this process to avoid the creation of `gcontact-relation` alongside `contact-relation` seems to be collapsing `contact` and `gcontact` while extracting the user-specific fields in `contact` to `user-contact`. No matter what we do or don't to `gfollower` and `contact-relation` which can be a second step.

Reply:
Disk space is some issue. So my plans are now the following:
* not implementing `gcontact-relation`
* removing `gfollower`
* implementing `follower` (like `gfollower` - but based on the contacts)
* implementing a functionality to create a `contact` entry without the proxied picture paths
* implementing a function to create these proxied paths when the contact is used for an item. (Or it is a personal contact)

if there is a need for a table that is a mxiture between `follower` and `contact-relation` we can always implement it as a view. Having a single table with several "type" fields indicating the different purposes of the table is a bad style in my eyes. We had enough of this in the past (see for example the old "term" table), we shouldn't repeat this error in the future.

Concerning this "user-contact" thing: This is a long term idea. At first the "post" stucture needs to be implemented, we gain much more of this.

Reply:
I do not think we need a `type` field that I loathe as well. For me, relations and interactions are both unique directional information from a contact to another. "X follows Y" is the same type of information as "X last interacted on Y's content". So the unifying `contact-relation` table would look like this:

Field | Type | Index | Extra
------|------|-------|------
actor_id | int(11) | PRIMARY | Foreign key to `contact.id`
target_id | int(11) | PRIMARY | Foreign key to `contact.id`
follows | tinyin(1) | | Boolean value
last-interaction | datetime | |

That's it. We can even throw a `timestamp` column to record the last time the `follows` value was changed if necessary.

Thank you for considering my plea.

Reply:
For this `follows` field we will need some `updated` field as well, since we have to see how current the data is. How would you name this table?

Reply:
I've started using the term `account` in my head to avoid any collision with the current uses of `contact` which are multiple and limited.

So we would have:
- `account`: public contact information enabling local profile page for remote contacts. This includes contact information about local users (what is currently `self=1`).
- `user-account`: user-specific information related to an account. This would include user blocks and ignores, contact notes and pull frequency, but not follow status.
- `account-relation`: This would hold the follow status between accounts, including local users through their public account. It also holds the last interaction time.

The only thing I'm not completely happy about is that we would lose the immediate information about two-ways relationships since we would have to join the `account-relation` table twice to determine both sides of the relationship. It would use indexes, but still.

Reply:
The relation table does not replace the fields in the contact table that are used to define the status like "friend". This is the represantation of the published relations that can be found in the corresponding Ap endpoints. So when someone - for example - hides someones relation, then this relation won't appear here. That still would be something for the `user-account` table.

Reply:
I'm not sure I follow? Hiding the relationship would be in `user-account` but the relationship itself would still be in `account-relation`.

Reply:
No. This table is meant for seeing the relations between contacts - mostly for contacts that aren't on your system. If we would replace our existing system with that, then we would need several additional fild in that table (like the id-values for DFRN, the individual key for DFRN and the id value of the contact request from AP. It would make no sense at all to have fields for all this stuff in this table. This will be - at some time - part of the `account-user` table.

BTW: I realized that we shouldn't call this table `account-relation` at the moment, since we don't have an account table and won't have in the next months. So I would extend the existing `contact-relation` table in the moment. This table would create some foreign key to the `contact` table at the moment. This couldn't easily be switched to the account table.

Reply:
Thanks for the elaboration about relationships with local users. I agree this table wouldn't be the good place for this kind of information, but the en result (follower, followed or friends) should still appear in the global `account-relation` table.

> This couldn't easily be switched to the account table.

Could it or could it not?

Reply:
It could not be switched. We will have different account ids than contact ids. So we need to copy this data to some temporary table, then clear that table, then we had to remove the foreign keys, add the new foreign keys and fill the table again. Or we use `contact-relation` and use some easy query to fill `account-relation` from it - without any fear of losing thsat when something goes wrong.

Reply:
I'd prefer that last option, hence the new model name I started using internally so that we can leave the existing structure while we build the improved one.